vsutil.py

```diff
@@ -157,12 +162,59 @@ def get_w(height: int, aspect_ratio: float = 16 / 9, only_even: bool = True) ->
     """
     width = height * aspect_ratio
     if only_even:
-        width = round(width / 2) * 2
-    return int(round(width))
+        return round(width / 2) * 2
+    return round(width)
 
 
 def is_image(filename: str) -> bool:
     """
     Returns true if a filename refers to an image.
     """
-    return mimetypes.types_map.get(os.path.splitext(filename)[-1], "").startswith("image/")
+    return types_map.get(path.splitext(filename)[-1], "").startswith("image/")
+
+
+def resample(clip: vs.VideoNode, bitdepth: int, *, force_float: bool = False) -> vs.VideoNode:
```

Comment:
This function feels very verbose. Does it do more than just

```py
def depth(clip: vs.VideoNode, bitdepth: int, *, force_float: bool = False) -> vs.VideoNode:
    """
    A bit depth converter only using core.resize and the PresetFormats defined in vscore.cpp.
    By default, will always output INTEGER sample type.
    :param bitdepth: Desired bits_per_sample of output clip.
    :param force_float: Forces output clip to have FLOAT sample type,
        only really useful for 16/32 bit conversions where the color family
        supports both INTEGER and FLOAT at these bit depths.
    :return: Converted clip with desired bit depth. ColorFamily will be same as input.
    """
    curr_depth = get_depth(clip)
    target_sample_type = vs.FLOAT if force_float else vs.INTEGER
    # nothing to do
    if curr_depth == bitdepth and clip.format.sample_type = target_sample_type:
        return clip
    dither = 'error_diffusion' if curr_depth > bitdepth else dither = 'none'

    return clip.resize.Point(format=clip.format.replace(bits_per_sample=bitdepth, sample_type=target_sample_type), dither_type=dither)
```
?
I also prefer the name depth or something that very clearly communicates that this is just depth conversion.
And I think we should default to float for 32bit output because 32bit int is very rarely what a user wants or expects.

Reply:
Was discussed on Discord: since switching either sample_type or bits_per_sample will add an overhead, even if one of them is the same as the source clip, it *can* be reduced to the following with no effect on the tests (or performance theoretically):
```py
def resample(clip: vs.VideoNode, bitdepth: int, *, sample_type: Optional[Literal[vs.FLOAT, vs.INTEGER]] = None) \
        -> vs.VideoNode:
    """
    A bit depth converter only using core.resize and Format.replace.
    By default, outputs FLOAT sample type for 32 bit and INTEGER for anything else.

    :param bitdepth:    Desired bits_per_sample of output clip.
    :param sample_type: Desired sample_type of output clip. Allows overwriting default FLOAT/INTEGER behavior.
    :return: Converted clip with desired bit depth and sample type. ColorFamily will be same as input.
    """
    curr_depth = get_depth(clip)
    if not sample_type:
        if bitdepth == 32:
            sample_type = vs.FLOAT
        else:
            sample_type = vs.INTEGER

    if curr_depth == bitdepth and clip.format.sample_type == sample_type:
        return clip

    dither = 'error_diffusion' if curr_depth > bitdepth and sample_type == vs.INTEGER else 'none'

    return clip.resize.Point(format=clip.format.replace(bits_per_sample=bitdepth, sample_type=sample_type),
                             dither_type=dither)
```

Reply:
Please check for `if sample_type is None` explicitly. The sample type is an enum, and vs.INTEGER is 0, so
```py
>>> bool(vs.INTEGER)
False
```

Reply:
Oh nice catch, fixed:
```py
def resample(clip: vs.VideoNode, bitdepth: int, *, sample_type: Optional[Literal[vs.FLOAT, vs.INTEGER]] = None) \
        -> vs.VideoNode:
    """
    A bit depth converter only using core.resize and Format.replace.
    By default, outputs FLOAT sample type for 32 bit and INTEGER for anything else.

    :param bitdepth:    Desired bits_per_sample of output clip.
    :param sample_type: Desired sample_type of output clip. Allows overriding default FLOAT/INTEGER behavior.
    :return: Converted clip with desired bit depth and sample type. ColorFamily will be same as input.
    """
    curr_depth = get_depth(clip)
    if sample_type is None:
        if bitdepth == 32:
            sample_type = vs.FLOAT
        else:
            sample_type = vs.INTEGER

    if curr_depth == bitdepth and clip.format.sample_type == sample_type:
        return clip

    dither = 'error_diffusion' if curr_depth > bitdepth and sample_type == vs.INTEGER else 'none'

    return clip.resize.Point(format=clip.format.replace(bits_per_sample=bitdepth, sample_type=sample_type),
                             dither_type=dither)
```

Reply:
- There should be a way to set the dither_type, in many situations you don't need dithering.

- Also, theoretically, dithering has to be done when upsampling full range content. (btw, zimg will not do dithering when upsampling limited range int to limited range int, even if specified.)

- When going float --> int there is currently no way to output full range.

- I would use a different name, since resampling can be much more than just bitdepth resampling.

Reply:
Good point on the `dither_type` param, definitely should be changeable :+1: , not familiar enough with the zimg codebase nor vs to comment on the 2nd and 3rd point. Regarding the 4th point, the function name isn't set in stone yet, but what other resampling can be done besides changing the `sample_type` and the `bits_per_sample` as this currently lets you change both.

Reply:
>Regarding the 4th point, the function name isn't set in stone yet, but what other resampling can be done besides changing the sample_type and the bits_per_sample as this currently lets you change both.

e. g. size, subsampling, color format, chroma alignment
All of those can be changed with `fmtc.resample`.
`depth` might be an appropriate name, since you basically only want to call this function when you want to change the bitdepth. But maybe you have a better idea. There is also `placebo.Resample` which just lets you change the size. I'd find it counterintuitive for users if there are multiple "resample" functions with different features.

Regarding the second point. I think you can keep this default:
```python
    dither = 'error_diffusion' if curr_depth > bitdepth and sample_type == vs.INTEGER else 'none'
```
But I am not a 100% certain.
Strictly speaking, you need dithering when raising the bitdepth of full range input, or when converting between full and limited range. But it probably doesn't matter at high bitdepths and not using dithering is _much_ faster.